doc: add annotation to writeFile `data` as `Object` (#39167)
nodejs/node: doc/api/fs.md

Write `buffer` to the file.

> If `buffer` is a plain object, like all variations of `writeFile`, it must have

Reviewer: Can you remove this bit? I don't think it brings much value and makes the sentence more complicated than it needs to be :)

```suggestion
If `buffer` is a plain object, it must have
```

Author: Sure. I originally included it for consistency

Reviewer: I meant, can you remove it everywhere? I agree we should aim for consistency.

Author: Oh 🤦‍♂️ sure, yes, will do tonight

-->

* `file` {string|Buffer|URL|FileHandle} filename or `FileHandle`
* `data` {string|Buffer|TypedArray|DataView|AsyncIterable|Iterable|Stream|

> Object}*

Author: Re-ordered so `Object` is last, and add `*`

Reviewer: The `*` doesn't refer to anything until 2500 lines below?

Reviewer 2: Sorry, I didn't mean a literal asterisk. I would just add the language used in https://github.com/nodejs/node/pull/34993 only to the parts where only the metadata and data types were updated and whose description doesn't consist of a documentation "redirect".

The only API I can see that seems to be missing the wording and doesn't have a documentation "redirect" is `filehandle.write(buffer[, offset[, length[, position]]])`.

Reviewer 2: Also I think the "(not inherited)" part is redundant, as that's implied with an "own property".

Author: > Also I think the "(not inherited)" part is redundant, as that's implied with an "own property".

Yes, intentionally redundant though: I added it to specifically call out / emphasis the `toString()` cannot come from where the user would reasonably expect. I feel like with gotchas (and this is a very bizarre gotcha), they should be abundantly clear.

Reviewer 2: Then please make sure to update all object `toString()` references in this document for consistency.

> * `data` {string|Buffer|TypedArray|DataView|AsyncIterable|Iterable|Stream|
> Object}*

Reviewer: You need to keep this on one line, otherwise the parser doesn't recognize this as a list of types
```suggestion
* `data` {string|Buffer|TypedArray|DataView|AsyncIterable|Iterable|Stream|Object}*
```

Author: It was already not 1 line; I merely swapped what was on the second line.

Your suggestion violates the 80 character line length limit :(

Reviewer: My bad, this should work:
```suggestion
* `data` {string|Buffer|TypedArray|DataView|AsyncIterable|Iterable|Stream
  |Object} *
```

a file descriptor.

The `encoding` option is ignored if `data` is a buffer.

> \* If `data` is a normal object, it must have an own (not inherited) `toString`
> function property.

Reviewer: Avoid `normal object`; my suggestion is not ideal so if you have a better suggestion please go ahead.
```suggestion
\* If `data` is of type `object`, but is not an {ArrayBufferView} and it doesn't
 have an own `toString` property, a `TypeError` is thrown.
```

Author: I didn't add "normal object". I think "plain object" is probably good though.

Author: Cool, will update in my next commit 🙂

-->

* `file` {string|Buffer|URL|integer} filename or file descriptor

> * `data` {string|Buffer|TypedArray|DataView|Object}*

Reviewer: The `*` refer to the paragraph 1000 lines above?

* `encoding` {string}
* Returns: {number} The number of bytes written.

> If `string` is a plain object, like all variations of `writeFile`, it must have

Author: I'm not sure if this is applicable, but `Object` is listed in the types, so I guessed yes. Wanted to specifically call out just in case.
